### PR TITLE
Added endpoints for pool stats and pool leaderboard

### LINKED ISSUE #225: [Backend] Hype Pool leaderboard API endpoint
Returns top X Telegram accounts sorted by their total number of impressions.

We also need to receive the TG username from the analytics layer.

### PR DESCRIPTION
**Pull request by Malak67**
# Generic Pull Request template

## Linked issue

Please link the issue:

ex. Closes #225 #224

## Checklist before requesting a review

- [x] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests.
- [ ] Do we need to implement analytics?
- [ ] Will this be part of a product update? If yes, please write one phrase about this update.
- [ ] The changes impact front-end.

## For changes impacting frontend

Please leave some screenshots/video recordings of the newly implemented interface and the Figma link for its initial design.

### Screenshots/ recordings

=> Multimedia here.
image
image

### Figma link

=> Figma link here.

## Description

Please provide a short description about what the PR contains.

## Funny gif

Leave a funny gif for faster review.

### CODE REVIEW

File: services/hype-backend/src/modules/reward/reward.service.ts [hunk 1]
```diff
@@ -308,4 +314,55 @@ export class RewardService {
...
+  async getPoolStats(poolId: string): Promise<PoolStatsResult> {
+    const claims = await this.claimRepository.find({ where: { poolId } });
+
+    let tokensAwarded = 0;
+    let tokensClaimed = 0;
+
+    for (const claim of claims) {
+      const amount = parseFloat(claim.amount);
+      if (claim.claimed) {
+        tokensAwarded += amount;
```
Comment (VargaElod23) [inline services/hype-backend/src/modules/reward/reward.service.ts:327]: Shouldn't this be reversed? if claimed add to claimed

**Review by VargaElod23 — CHANGES_REQUESTED** [review]
Single logic question added

File: services/hype-backend/src/modules/reward/reward.service.ts (hunk 1 above)
Comment (Malak67) [reply services/hype-backend/src/modules/reward/reward.service.ts:327]: This is true for the reward entity. But we are using the claim entity. The field claimed = true means the user has claimed from the contract and got the reward.

File: services/hype-backend/src/modules/reward/pools.controller.ts [hunk 2]
```diff
@@ -34,12 +39,38 @@ export class PoolsController {
...
-    description: 'Returns joined pools for a givven address',
+    description: 'Returns joined pools for a given address',
   })
   async getJoined(@Param('address') address: string): Promise<IPool[]> {
     return this.rewardService.getJoinedPools(address);
   }
 
+  @Get('stats/:poolId')
+  @UseGuards(WalletGuard)
+  @ApiBearerAuth('authorization')
```
Comment (rattrap) [inline services/hype-backend/src/modules/reward/pools.controller.ts:50]: Don't need auth for this.

File: services/hype-backend/src/modules/reward/pools.controller.ts [hunk 3]
```diff
@@ -34,12 +39,38 @@ export class PoolsController {
...
+    type: PoolStatsDto,
+    description: 'Returns pool stats for a given address',
+  })
+  async getStats(@Param('poolId') poolId: string): Promise<PoolStatsResult> {
+    return this.rewardService.getPoolStats(poolId);
+  }
+
+  @Get('leaderboard/:poolId')
+  @UseGuards(WalletGuard)
+  @ApiBearerAuth('authorization')
```
Comment (rattrap) [inline services/hype-backend/src/modules/reward/pools.controller.ts:62]: Don't need auth for this.

File: services/hype-backend/src/modules/reward/reward.service.ts [hunk 4]
```diff
@@ -33,6 +34,13 @@ export interface ClaimResult {
   claim: HypeClaim;
 }
 
+export interface PoolStatsResult {
+  tokensAwarded: number;
+  tokensClaimed: number;
+  participants: number;
+  impressions: number;
+}
```
Comment (rattrap) [inline services/hype-backend/src/modules/reward/reward.service.ts:42]: Why do need this? We can just use the DTO

File: services/hype-backend/src/modules/reward/reward.service.ts [hunk 5]
```diff
@@ -308,4 +314,52 @@ export class RewardService {
     );
     return pools;
   }
+
+  async getPoolStats(poolId: string): Promise<PoolStatsResult> {
+    const claims = await this.claimRepository.find({ where: { poolId } });
```
Comment (rattrap) [inline services/hype-backend/src/modules/reward/reward.service.ts:319]: This is wrong. We should look in `rewards`, not claims

File: services/hype-backend/src/modules/reward/reward.service.ts [hunk 6]
```diff
@@ -308,4 +314,52 @@ export class RewardService {
...
+    for (const reward of rewards) {
+      impressions += parseFloat(reward.impressions.toString());
+    }
+
+    return {
+      tokensAwarded,
+      tokensClaimed,
+      participants,
+      impressions,
+    };
```
Comment (rattrap) [inline services/hype-backend/src/modules/reward/reward.service.ts:346]: Rewards are in the thousands at least. We shouldn't interate over them.
Let's just do two queries:
- distinct telegram id count
- sum(impressions)

**Review by rattrap — CHANGES_REQUESTED** [review]

**Review by rattrap — APPROVED** [review]

**Pull request merged**